pkg/webhook/admission/defaulter_custom.go

```diff
@@ -76,7 +88,14 @@ func (h *defaulterForType) Handle(ctx context.Context, req Request) Response {
 		return Errored(http.StatusBadRequest, err)
 	}
 
-	// Default the object
+	return h.handler.Handle(ctx, req.Object.Raw, obj)
```

Comment:
If we're going to introduce another way to implement mutation, I wonder if we should go all the way:
* Should we just hand over ctx and the entire request? (~ l.83)
* Should we call this interface Mutator & the builder func WithMutator

I think then it would also work well for resource updates

Reply:
@vincepri @alvaroaleman Opinions?

Reply:
I wanted to avoid doing `h.decoder.Decode(req, obj)` every time, but sure, better leave everything up to the developer.

Reply:
Sorry, I don't quite follow how the new thing helps with the problem at hand. The problem at hand is you need the raw request or not? Wouldn't it be better to change the existing `CustomDefaulter` interface to add it as an arg in there?

Reply:
Yes, I do need the raw request. But this is not the biggest problem, because the request is embedded in the context.

But the reason why I need the raw request is to be able to return json patches. And the CustomDefaulter currently just expects changes to the object. The creation of patches is completely hidden from the defaulter.

We could change the interface to return patches, but that's a breaking change. Isn't it better to add a fully manual mechanism to building webhook handlers?

Reply:
My thinking was that we can lose information during json.Unmarshal and json.Marshal (either if the CRDs are not prefectly designed with omitempty, or if we have to wrong version of the CRD because it's not ours)

So I thought if we provide a low-level way to handle this better, let's go all the way and give folks all the info (req) and allow them to return the patches directly

Reply:
Somewhat related issue https://github.com/kubernetes-sigs/controller-runtime/issues/2800

Reply:
Would this work: https://github.com/kubernetes-sigs/controller-runtime/issues/2800#issuecomment-2381413954 ?

Reply:
Answered on the issue, nope